cmd/serve/restic/restic.go

```diff
@@ -0,0 +1,392 @@
+	}
+	w.Header().Set("Content-Length", strconv.FormatInt(size, 10))
+
+	file, err := o.Open(options...)
+	if err != nil {
+		fs.Debugf(remote, "Get request open error: %v", err)
+		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		return
+	}
+	defer func() {
+		err := file.Close()
+		if err != nil {
+			fs.Errorf(remote, "Get request: close failed: %v", err)
+		}
+	}()
+
+	w.WriteHeader(code)
+
+	_, err = io.Copy(w, file)
+	if err != nil {
+		fs.Errorf(remote, "Didn't finish writing GET request: %v", err)
+		return
+	}
+}
+
+// postObject posts an object to the repository
+func (s *server) postObject(w http.ResponseWriter, r *http.Request, remote string) {
+	// fs.Debugf(s.f, "content length = %d", r.ContentLength)
+	if r.ContentLength >= 0 {
+		// Size known use Put
+		info := object.NewStaticObjectInfo(remote, time.Now(), r.ContentLength, true, nil, s.f)
+		_, err := s.f.Put(r.Body, info)
+		if err != nil {
+			fs.Errorf(remote, "Post request put error: %v", err)
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+	} else {
+		// Size unknown use Rcat
+		_, err := operations.Rcat(s.f, remote, r.Body, time.Now())
+		if err != nil {
+			fs.Errorf(remote, "Post request rcat error: %v", err)
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+// delete the remote
+func (s *server) deleteObject(w http.ResponseWriter, r *http.Request, remote string) {
+	o, err := s.f.NewObject(remote)
+	if err != nil {
+		fs.Debugf(remote, "Delete request error: %v", err)
+		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		return
+	}
+
+	if err := o.Remove(); err != nil {
+		fs.Errorf(remote, "Delete request remove error: %v", err)
+		if err == fs.ErrorObjectNotFound {
+			http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		} else {
+			http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+		}
+		return
+	}
+}
+
+// listItem is an element returned for the restic v2 list response
+type listItem struct {
+	Name string `json:"name"`
+	Size int64  `json:"size"`
+}
+
+// return type for list
+type listItems []listItem
+
+// add a DirEntry to the listItems
+func (ls *listItems) add(entry fs.DirEntry) {
+	if o, ok := entry.(fs.Object); ok {
+		*ls = append(*ls, listItem{
+			Name: path.Base(o.Remote()),
+			Size: o.Size(),
+		})
+	}
+}
+
+// listObjects lists all Objects of a given type in an arbitrary order.
+func (s *server) listObjects(w http.ResponseWriter, r *http.Request, remote string) {
+	fs.Debugf(remote, "list request")
+
+	var ls listItems
+	err := walk.Walk(s.f, remote, true, -1, func(path string, entries fs.DirEntries, err error) error {
+		if err == nil {
+			for _, entry := range entries {
+				ls.add(entry)
+			}
+		}
+		return err
+	})
+	if err != nil {
+		fs.Errorf(remote, "list failed: %v", err)
```

Comment:
This needs handling non-existing dirs, e.g.:
```diff
diff --git a/cmd/serve/restic/restic.go b/cmd/serve/restic/restic.go
index f5c56b3a..4df6a3b0 100644
--- a/cmd/serve/restic/restic.go
+++ b/cmd/serve/restic/restic.go
@@ -17,6 +17,7 @@ import (
    "github.com/ncw/rclone/cmd/serve/httplib"
    "github.com/ncw/rclone/cmd/serve/httplib/httpflags"
    "github.com/ncw/rclone/fs"
+   "github.com/ncw/rclone/fs/fserrors"
    "github.com/ncw/rclone/fs/object"
    "github.com/ncw/rclone/fs/operations"
    "github.com/ncw/rclone/fs/walk"
@@ -362,10 +363,14 @@ func (s *server) listObjects(w http.ResponseWriter, r *http.Request, remote stri
        }
        return err
    })
+
    if err != nil {
-       fs.Errorf(remote, "list failed: %v", err)
-       http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
-       return
+       _, err = fserrors.Cause(err)
+       if err != fs.ErrorDirNotFound {
+           fs.Errorf(remote, "list failed: %#v %T", err, err)
+           http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+           return
+       }
    }
```

Otherwise it fails e.g. for a non-existing local repo.